freestyle-cache/shared/src/main/scala/cache.scala

```diff
@@ -0,0 +1,117 @@
+package freestyle
+
+import cats.{~>}
+
+package cache {
+
+  class KeyValueProvider[Key, Val] {
+
+    /* CacheM trait is a type-class of functors for which key-value store operations
+     * can be provided.
+     *
+     * We assume that the actual store is too big or remote to allow for general
+     * operations over all values, or to search or to iterate over all keys
+     */
+    @free sealed trait CacheM[F[_]] {
+
+      // Gets the value associated to a key, if there is one */
+      def get(key: Key): FreeS.Par[F, Option[Val]]
+
+      // Sets the value of a key to a newValue.
+      def put(key: Key, newVal: Val): FreeS.Par[F, Unit]
```

Comment:
I would add the following methods if we can support them both on the InMemory and Redis impl.
```scala
def putAll(keyValues: Map[Key, Val]): FreeS.Par[F, Unit]
def putIfAbsent(key: Key, newVal: Val): FreeS.Par[F, Unit]
def replace(key: Key, newVal: Val): FreeS.Par[F, Unit]
def isEmpty: FreeS.Par[F, Boolean]
def containsKey(key: Key): FreeS.Par[F, Boolean]
def containsValue(v: Val): FreeS.Par[F, Boolean]
```